### DISCUSSION TITLE
How to replace dashboard data sources

### DISCUSSION THREAD

**Discussion by vishwa-ims** [RevealBi/Reveal.Sdk#33] [ASP.NET Core] [answered]
### Company

IMSPOS

### SDK Version

1.4

### Client Framework

ASP.NET Web App

### Server Platform

ASP.NET

### Description

We have below usecase

- We have existing dashboard with datasource "ds1"
- We downloaded those dashboard and uploaded back again with new datasource "ds2"
- But still new dashboard is looking for old datasource "ds1"

We wanted to understand how to update datasource of existing dashboards. We have number of dashboards for which we want to update datasource to point to new IP address.

**Comment by brianlagunas** [comment]
> We downloaded those dashboard and uploaded back again with new datasource "ds2"

How are you changing the data sources?

**Comment by vishwa-ims** [comment]
Below is a codebase for dashboard upload. We are passing sourceid of new source

if (!Directory.Exists(newpath))
            {
                Directory.CreateDirectory(newpath);
            }
            fileName = DashboardName + "-" + companyId + "-" + randomId + ".rdash";
            pathMerg = Path.Combine(newpath, fileName);
            using (var stream = new FileStream(pathMerg, FileMode.Create))
            {
                files.CopyTo(stream);
            }

            /////////////Add new Dashboard///////////////

            Dashboards dashboards = new Dashboards();
            dashboards.CompanyId = Convert.ToInt32(companyId);
            dashboards.DataSourceId = Convert.ToInt32(sourceId);
            dashboards.DashboardId = randomId.ToString();
            dashboards.DashboardName = DashboardName + "-" + companyId + "-" + randomId;
            dashboards.DisplayDashboardName = DashboardName;
            dashboards.DashboardPath = pathMerg;
            dashboards.IsActive = true;
            await _dashboardService.CreateDashboard(dashboards);

        

            //////////////////////Make Entry on DashboardUser Table//////////////////////////
            DashboardUser dashboardUser = new DashboardUser();
            dashboardUser.DashboardId = Convert.ToInt32(dashboards.DashboardId);
            dashboardUser.UserId = Convert.ToInt32(HttpContext.Session.GetString("UserId"));
            dashboardUser.PermissionType = Convert.ToInt32(Stadis.Intelligence.Data.Enum.Enum.PermissionType.Owner);
            dashboardUser.SourceType = Convert.ToInt32(Stadis.Intelligence.Data.Enum.Enum.SourceType.Dashboard);
            await _dashboardService.AddAssigneesToDashboard(dashboardUser);

            /////////////Add new Dashboard in Folder Table///////////////
            DashboardFolder dashboardFolder = new DashboardFolder();
            dashboardFolder.DashboardId = Convert.ToInt32(dashboards.DashboardId);
            dashboardFolder.FolderId = Convert.ToInt32(HttpContext.Session.GetString("FolderId"));
            await _dashboardService.SaveDashboardInFolder(dashboardFolder);

**Comment by brianlagunas** [comment]
Unfortunately, that code does not show the actual replacement of data sources in the dashboard or rdash file. That looks like custom code.

How are you replacing the data sources of the dashboard?

**Comment by brianlagunas** [comment]
I converted this to a discussion since this is not a bug.

**Reply by vishwa-ims** [reply]
@brianlagunas  I have basic question . If a dashboard is having multiple visualizations with different data source for each visualization. If we export the dashboard and import again , can we assign new datasource for all these visulizations independently ?

**Reply by brianlagunas** [reply, marked as answer]
There is no built-in API to change data sources of a dashboard. Currently, you would have to use the RevealView editor to manually change the data source of each visualization.

Having said that, the rdash file is just a json file. If you change the rdash extension to a zip, you can see the json file that represents the dashboard. You could modify the json directly, but that would be an unsupported approach and you should use at your own risk. We have other customers taking this approach to modify their data sources.

**Reply by vishwa-ims** [reply]
Thanks @brianlagunas . Could you please help me in understanding what is RevealView editor and how I can use it ? If you can share any reference link that will be really helpful . Thank you

**Reply by brianlagunas** [reply]
The RevealView editor is this screen: 
![image](https://github.com/RevealBi/Reveal.Sdk/assets/835562/24bc125a-a5d5-4685-bf4d-baefac61da96)

![image](https://github.com/RevealBi/Reveal.Sdk/assets/835562/d6c1aad9-74fe-4dd3-a82f-30d736509443)

![image](https://github.com/RevealBi/Reveal.Sdk/assets/835562/b6a8b9e7-0a7c-4c23-ac35-d9e80e980025)

**Reply by vishwa-ims** [reply]
Hi @brianlagunas  I tried to manually edit rdash file . But it is giving me error message as "Data source not found : 71" though datasource with id 71 is available in my sql server database. Is there anything I am missing ?

**Reply by brianlagunas** [reply]
It is difficult to answer that without a reproducible dashboard. @TorreyBetts or @luispandolfi can you think of anything that would cause an error like that when trying to change data sources?

**Reply by vishwa-ims** [reply]
I was able to fix it by modifying rdash json file . Thank you @brianlagunas  for your support
